server/config.py

```diff
@@ -26,23 +28,40 @@
 KEY_SECURE_SESSION_COOKIES = 'secure_session_cookies'
 KEY_SAMESITE_SESSION_COOKIES = 'samesite_session_cookies'
 
+from distutils.util import strtobool
 
 #
 # Container for properties that should be read on startup, and maybe be
 # refreshed dynamically.
 #
-class Config:
+# If using redis all the keys are cached in the redis server, otherwise
+# the keys are just the attributes of this class.  Note that redis does
+# not support boolean values.  Hence care must be taken with the string
+# conversion.
+#
+class Config(dict):
+
+    def __init__(self):
+        if constants.REDIS_IP_ADDR is not None:
+            self.red = redis.Redis(constants.REDIS_IP_ADDR, port=6379)
+
+    def __getitem__(self, key):
+        if constants.REDIS_IP_ADDR is not None:
+            # TODO: Rethink how this is done as it will break if we store anything in the config other than a boolean.
+            return True if strtobool(str(self.red.get(key), 'utf-8')) else False
+        else:
+            return super(Config, self).__getitem__(key)
 
-    training_enabled = True
-    use_tpu = True
-    secure_session_cookies = True
-    samesite_session_cookies = True
+    def __setitem__(self, key, value):
+       if constants.REDIS_IP_ADDR is not None:
+           self.red.set(key, str(value))
+       super(Config, self).__setitem__(key, value)
 
     def reset(self):
-        self.training_enabled = True
-        self.use_tpu = True
-        self.secure_session_cookies = True
-        self.samesite_session_cookies = True
+        self[KEY_TRAINING_ENABLED] = True
+        self[KEY_USE_TPU] = True
+        self[KEY_SECURE_SESSION_COOKIES] = True
+        self[KEY_SAMESITE_SESSION_COOKIES] = True
 
     def refresh(self):
         client = datastore.Client()
```

Comment:
I'm confused. We still read from the datastore?

Reply:
Yes.  Redis does not replace the datastore, it just acts as a low latency, memory store for distribution of the configuration items among multiple instances.  Each instance, however, upon startup, will read the datastore and "refresh" the redis instance in the case that a configuration item had changed.

As discussed out of band, the initial thought was that the config items would be very, very static and doing reads from the datastore each time an item was accessed was overkill.  Hence the "read once at startup" implementation.  Things always get more complicated when data needs to be distributed, but I think we don't really care when the "refreshes" happen, or if a "refresh" happens when it doesn't really need to.